File: backend/app/schemas/content.py
from pydantic import BaseModel, field_validator
from typing import Optional, Dict, List, Any


class ContentItemBase(BaseModel):
    """Base content item schema"""
    title: str
    topic: str
    subtopic: Optional[str] = None
    difficulty_level: str  # easy, normal, hard, challenge
    format: str  # text, visual, video, interactive
    content_type: str  # lesson, exercise, quiz, explanation
    content_data: Dict[str, Any]
    reference_answer: Optional[Dict[str, Any]] = None
    hints: List[Any] = []
    explanations: List[Any] = []
    skills: List[str] = []
    prerequisites: List[str] = []
    extra_data: Dict[str, Any] = {}

    @field_validator('difficulty_level')
    @classmethod
    def difficulty_level_valid(cls, v):
        allowed_levels = ['easy', 'normal', 'hard', 'challenge']
        if v not in allowed_levels:
            raise ValueError(f'Difficulty level must be one of: {", ".join(allowed_levels)}')
        return v

    @field_validator('format')
    @classmethod
    def format_valid(cls, v):
        allowed_formats = ['text', 'visual', 'video', 'interactive']
        if v not in allowed_formats:
            raise ValueError(f'Format must be one of: {", ".join(allowed_formats)}')
        return v

    @field_validator('content_type')
    @classmethod
    def content_type_valid(cls, v):
        allowed_types = ['lesson', 'exercise', 'quiz', 'explanation']
        if v not in allowed_types:
            raise ValueError(f'Content type must be one of: {", ".join(allowed_types)}')
        return v

    @field_validator('title', 'topic')
    @classmethod
    def text_not_empty(cls, v):
        if not v or not v.strip():
            raise ValueError('Field cannot be empty')
        return v


class ContentItemCreate(ContentItemBase):
    """Schema for creating a content item"""
    pass


class ContentItemUpdate(BaseModel):
    """Schema for updating a content item"""
    title: Optional[str] = None
    topic: Optional[str] = None
    subtopic: Optional[str] = None
    difficulty_level: Optional[str] = None
    format: Optional[str] = None
    content_type: Optional[str] = None
    content_data: Optional[Dict[str, Any]] = None
    reference_answer: Optional[Dict[str, Any]] = None
    hints: Optional[List[Any]] = None
    explanations: Optional[List[Any]] = None
    skills: Optional[List[str]] = None
    prerequisites: Optional[List[str]] = None
    extra_data: Optional[Dict[str, Any]] = None

    @field_validator('difficulty_level')
    @classmethod
    def difficulty_level_valid(cls, v):
        if v is not None:
            allowed_levels = ['easy', 'normal', 'hard', 'challenge']
            if v not in allowed_levels:
                raise ValueError(f'Difficulty level must be one of: {", ".join(allowed_levels)}')
        return v

    @field_validator('format')
    @classmethod
    def format_valid(cls, v):
        if v is not None:
            allowed_formats = ['text', 'visual', 'video', 'interactive']
            if v not in allowed_formats:
                raise ValueError(f'Format must be one of: {", ".join(allowed_formats)}')
        return v

    @field_validator('content_type')
    @classmethod
    def content_type_valid(cls, v):
        if v is not None:
            allowed_types = ['lesson', 'exercise', 'quiz', 'explanation']
            if v not in allowed_types:
                raise ValueError(f'Content type must be one of: {", ".join(allowed_types)}')
        return v


class ContentItemResponse(ContentItemBase):
    """Schema for content item response"""
    content_id: int

    class Config:
        from_attributes = True


class PaginationMetadata(BaseModel):
    """Pagination metadata schema"""
    total: int
    limit: int
    offset: int
    total_pages: int
    current_page: int
    has_next: bool
    has_prev: bool


class ContentListResponse(BaseModel):
    """Response schema for paginated content list"""
    items: List[ContentItemResponse]
    pagination: PaginationMetadata

File: frontend/src/types/content.ts
/**
 * Content-related TypeScript types
 * Maps to backend schema: @backend/app/schemas/content.py
 * Backend contract source: @backend/scripts/content_generator.py
 */

/**
 * Difficulty level enumeration
 * @backend/app/schemas/content.py:21-27
 */
export type DifficultyLevel = 'easy' | 'normal' | 'hard' | 'challenge';

/**
 * Content format enumeration
 * @backend/app/schemas/content.py:29-35
 */
export type ContentFormat = 'text' | 'visual' | 'video' | 'interactive';

/**
 * Content type enumeration
 * @backend/app/schemas/content.py:37-43
 */
export type ContentType = 'lesson' | 'exercise' | 'quiz' | 'explanation';

/**
 * Lesson content_data shape (backend contract)
 * @backend/scripts/content_generator.py:get_lesson_content
 */
export interface LessonContentData {
  introduction: string;
  sections: Array<{ heading: string; content: string }>;
  key_points: string[];
}

/**
 * Exercise content_data shape (backend contract)
 * @backend/scripts/content_generator.py:get_exercise_content
 */
export interface ExerciseContentData {
  question: string;
  description?: string;
  starter_code?: string;
  solution?: string;
  test_cases?: string[];
}

/**
 * Quiz content_data shape (backend contract)
 * @backend/scripts/content_generator.py:get_quiz_content
 */
export interface QuizContentData {
  question: string;
  options: string[];
  correct_answer: string;
  explanation: string;
  quiz_type?: 'multiple_choice' | 'true_false' | 'multiple_select';
}

/**
 * Visual content_data shape
 */
export interface VisualContentData {
  image_url?: string;
  caption?: string;
  description?: string;
  alt_text?: string;
}

/**
 * Video content_data shape
 */
export interface VideoContentData {
  video_url?: string;
  duration?: number;
  subtitles?: string;
  thumbnail_url?: string;
}

/**
 * Reference answer structure for exercises
 */
export interface ReferenceAnswer {
  solution?: string;
  answer?: string;
  correct_answer?: string;
  value?: string;
}

/**
 * Content item interface matching backend ContentItemResponse schema
 * @backend/app/schemas/content.py:ContentItemResponse
 */
export interface ContentItem {
  content_id: number;
  title: string;
  topic: string;
  subtopic: string | null;
  difficulty_level: DifficultyLevel;
  format: ContentFormat;
  content_type: ContentType;
  content_data: Record<string, any>;
  reference_answer: ReferenceAnswer | string | null;
  hints: any[];
  explanations: any[];
  skills: string[];
  prerequisites: string[];
  extra_data: Record<string, any>;
}

/**
 * Typed content items with discriminated unions
 */
export type LessonContentItem = ContentItem & {
  content_type: 'lesson';
  content_data: LessonContentData;
};

export type ExerciseContentItem = ContentItem & {
  content_type: 'exercise';
  content_data: ExerciseContentData;
};

export type QuizContentItem = ContentItem & {
  content_type: 'quiz';
  content_data: QuizContentData;
};

/**
 * Pagination metadata interface
 * @backend/app/schemas/content.py:PaginationMetadata (lines 110-119)
 */
export interface PaginationMetadata {
  total: number;
  limit: number;
  offset: number;
  total_pages: number;
  current_page: number;
  has_next: boolean;
  has_prev: boolean;
}

/**
 * Content list response with pagination
 * @backend/app/schemas/content.py:ContentListResponse (lines 121-124)
 */
export interface ContentListResponse {
  items: ContentItem[];
  pagination: PaginationMetadata;
}
